### PR TITLE
Support for JBang within Camel

### PR DESCRIPTION
**Pull request by orpiske**
Example usage: 

[Earhquake Example](https://asciinema.org/a/423125)

[JMS Sink and Source](https://asciinema.org/a/423121)

### CODE REVIEW

**Comment by orpiske** [comment]
In addition to the community feedback, there's still a few things I want to do before consider this ready:
- Cleanup deps ... maybe using a bom pom
- Take a look at verbosity and see if it can be tuned
- Review packaging ... right now, there's none

**Comment by orpiske** [comment]
A few limitations:

* Native mode is not supported because dependencies cannot be resolved during build time

**Review by davsclaus — APPROVED** [review]
Awesome work, this looks really nice with a quick startup via jbang - keep hacking and have some fun

**Review by oscerd — APPROVED** [review]

**Comment by orpiske** [comment]
@davsclaus Thanks! 

Here's a fresh new feature ... searching for Kamelets via CLI

[asciicast](https://asciinema.org/a/423188)

**Comment by davsclaus** [comment]
Oh the search thingy with the embedded catalog is lovely

File: dsl/camel-jbang/CamelJbang.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+        System.out.printf("A new lock file was created on %s. Delete this file to stop running%n",
+            lockFile.getAbsolutePath());
+        lockFile.deleteOnExit();
+
+        return lockFile;
+    }
+}
+
+@Command(name = "search", description = "Run a Kamelet")
```
Comment (zhfeng) [inline dsl/camel-jbang/CamelJbang.java:138]: A typo of description ? it should be "Search a Kamelet"

File: dsl/camel-jbang/CamelJbang.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+ */
+
+//DEPS info.picocli:picocli:4.5.0
+//DEPS org.apache.camel:camel-core:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-core-model:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-api:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-main:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-kamelet-main:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-file-watch:3.12.0-SNAPSHOT
+//DEPS org.apache.camel:camel-resourceresolver-github:3.12.0-SNAPSHOT
```
Comment (tadayosi) [inline dsl/camel-jbang/CamelJbang.java:26]: Yeah you can use BOM for the camel deps.
https://github.com/tadayosi/jbang-sandbox/blob/main/camel/camel.java#L2-L5

File: dsl/camel-jbang/CamelJbang.java (hunk 1 above)
Comment (orpiske) [reply dsl/camel-jbang/CamelJbang.java:138]: You are right! thanks @zhfeng :)

File: dsl/camel-jbang/CamelJbang.java (hunk 2 above)
Comment (orpiske) [reply dsl/camel-jbang/CamelJbang.java:26]: Thanks @tadayosi! I will try that one today :)

**Comment by orpiske** [comment]
As for packaging, I am not sure if it is actually needed. I was playing with Apache Short URL service and I think we could just have something like this on the website: 

```
jbang https://s.apache.org/try-camel-jbang-dev search --search-term=jms
``` 

Of course, we replace the `try-camel-jbang-dev` with `try-camel` or similar.

**Comment by maxandersen** [comment]
@orpiske add a jbang-catalog.json file and you can do jbang camel@apache/camel and it will show up on jbang.dev/appstore too :)

**Comment by orpiske** [comment]
> @orpiske add a jbang-catalog.json file and you can do jbang camel@apache/camel and it will show up on jbang.dev/appstore too :)

Cool, that makes it very nice! I'll include that :)

**Comment by maxandersen** [comment]
oh and jbang app install camel@apache/camel and you get camel in your path.

File: dsl/camel-jbang/CamelJbang.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            readFileByLine(indexFile, this::checkForMatches);
+        } else {
+            readFileByLine(indexFile, this::printAll);
+        }
+
+        return 0;
+    }
+}
+
+@Command(name = "CamelJbang", mixinStandardHelpOptions = true, version = "CamelJbang 3.12.0-SNAPSHOT",
```
Comment (maxandersen) [inline dsl/camel-jbang/CamelJbang.java:234]: fyi, its JBang, jbang is the command, Jbang is a mistake :)

Comment (orpiske) [reply dsl/camel-jbang/CamelJbang.java]: Thanks. I think I corrected it now :)

**Comment by davsclaus** [comment]
> @orpiske add a jbang-catalog.json file and you can do jbang camel@apache/camel and it will show up on jbang.dev/appstore too :)

Max, should the jbang-catalog.json file be in the root folder?

**Comment by orpiske** [comment]
I'm tempted to say that it does, but I guess @maxandersen can confirm for good. For now, I got it working with:

```
jbang app install CamelJBang@orpiske/camel/camel-16612~dsl/camel-jbang
```

**Comment by maxandersen** [comment]
If you want the short name then yes. 

It can also be in subfolders but then you need to add the path = longer name. 

Alternatively create Apache/jbang-catalog and it can be jbang camwl@apache - even shorter :)

**Comment by orpiske** [comment]
> If you want the short name then yes.
> 
> It can also be in subfolders but then you need to add the path = longer name.
> 
> Alternatively create Apache/jbang-catalog and it can be jbang camwl@apache - even shorter :)

@maxandersen Thanks. I tried moving the catalog file to the root and modified it like this: 

```
{
	"catalogs": {},
	"aliases": {
	  "CamelJBang": {
		"script-ref": "dsl/camel-jbang/CamelJBang.java",
		"description": "A JBang-based Camel app for running Kamelets"
	  }
	}
}
```

But it did not seem to have worked. Did I miss anything?

**Comment by maxandersen** [comment]
You'll need to use your branch repo name to have it work. 

The short name won't work until merged.

**Comment by orpiske** [comment]
> But it did not seem to have worked. Did I miss anything?

Nevermind. I had to remove `~/.jbang`. Now this worked:

`jbang app install CamelJBang@orpiske/camel/camel-16612`

**Comment by orpiske** [comment]
Thanks everyone for the reviews, comments and suggestions. I am marking this one as ready for review. 

I also updated the screen casts to match the latest fixes: 

Running Kamelet bindings:
[asciicast](https://asciinema.org/a/423346)

Searching for Kamelets in the catalog: 
[asciicast](https://asciinema.org/a/423349)

There are a few more things I'd like to do - like a command to create new Kamelets bindings and so on - but this requires changes in other sub-projects.

**Comment by maxandersen** [comment]
You shouldn't have to remove .Jbang. At most run with jbang --fresh to ensure you get the latest.

**Comment by orpiske** [comment]
I am currently working on removing the fixed camel version on the //DEPS part, so we don't have extra work releasing this. Therefore, I believe we can wait a bit more before merging it.

**Comment by davsclaus** [comment]
The more features you add, then the more we look forward to your write a blog post to introduce this and publish on the camel front page

**Comment by orpiske** [comment]
> The more features you add, then the more we look forward to your write a blog post to introduce this and publish on the camel front page

Oh, count me in @davsclaus :+1: I am looking forward to getting this one in a good shape and writing a nice blog post with that as well!

File: dsl/camel-jbang/CamelJBang.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            Resource resource = resolver.resolve("github:apache:camel-kamelets:docs/modules/ROOT/nav.adoc");
+
+            if (!resource.exists()) {
+                throw new Exception("The resource does not exist");
+            }
+
+            try (FileWriter w = new FileWriter(indexFile)) {
+                Reader resourceReader = resource.getReader();
+
+                resourceReader.transferTo(w);
```
Comment (orpiske) [inline dsl/camel-jbang/CamelJBang.java:151]: This API is Java 10+, so we cannot use it for the moment. I'll have to use a different one.

**Comment by orpiske** [comment]
Today's update is that I modified the packaging a bit so that: 

1. We don't rely on fixed dependency versions on the script 
2. While also ensuring the code remains valid Java/JBang code

The intention is to ensure it remains aligned with the dependencies on the Camel Core but still allowing for the nice JBang features such as app install and so on. This may be a bit unusual, so I described it in the module's `README.md` file, but at first sight it doesn't seem like it would impact the release process (might be good to review that, though).

**Comment by maxandersen** [comment]
In case you missed it jbang supports default property values. 

${camel.jbang.version:RELEASE} 

Then default use of jbang camel@apache/camel will get latest release. 

And you can always do jbang -Dcamel.jbang.version=3.0.2 camel@apache/camel for specific version control.

**Comment by orpiske** [comment]
> In case you missed it jbang supports default property values.
> 
> ${camel.jbang.version:RELEASE}

Thanks, I didn't know about the `:RELEASE` part!

> 
> Then default use of jbang camel@apache/camel will get latest release.
> 
> And you can always do jbang -Dcamel.jbang.version=3.0.2 camel@apache/camel for specific version control.

**Comment by orpiske** [comment]
Today's feature is that I added the ability to also search for component information (with links). 

The search command changed a bit, it is now `search kamelets --search-term=jms` or `search components --search-term=jms`. Later I'll add search for languages and data formats. 

Here's the screencast:
[asciicast](https://asciinema.org/a/424286)

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
Yeah sounds good, its good to get something out there and you can also write a blog , record a video or something to start introducing this

**Comment by orpiske** [comment]
> Yeah sounds good, its good to get something out there and you can also write a blog , record a video or something to start introducing this

Thanks. +1! Today I'll start writing the documentation to include on the website and I'll prepare a blog post with some screen casts or videos as well.

**Pull request merged**